renderdoc/driver/gl/gl_pixelhistory.cpp

```diff
@@ -277,13 +593,390 @@ void QueryFailedTests(WrappedOpenGL *driver, GLPixelHistoryResources &resources,
       }
     }
 
+    // ensure that history objects are one-to-one mapped with modEvent objects
+    RDCASSERT(history[i].eventId == modEvents[i].eventId);
     history[i].scissorClipped = failedTest == OpenGLTest::ScissorTest;
     history[i].stencilTestFailed = failedTest == OpenGLTest::StencilTest;
     history[i].depthTestFailed = failedTest == OpenGLTest::DepthTest;
     history[i].backfaceCulled = failedTest == OpenGLTest::FaceCulling;
   }
 }
-};
+
+void QueryShaderOutPerFragment(WrappedOpenGL *driver, GLReplay *replay,
+                               GLPixelHistoryResources &resources,
+                               const rdcarray<EventUsage> &modEvents, int x, int y,
+                               rdcarray<PixelModification> &history,
+                               const std::map<uint32_t, uint32_t> &eventFragments)
+{
+  driver->ReplayLog(0, modEvents[0].eventId, eReplay_WithoutDraw);
+  for(size_t i = 0; i < modEvents.size(); ++i)
+  {
+    auto it = eventFragments.find(modEvents[i].eventId);
+    uint32_t numFragments = (it != eventFragments.end()) ? it->second : 0;
+
+    if(numFragments <= 1)
+    {
+      if(i < modEvents.size() - 1)
+      {
+        driver->ReplayLog(modEvents[i].eventId, modEvents[i + 1].eventId, eReplay_WithoutDraw);
+      }
+      continue;
+    }
+
+    driver->glEnable(eGL_SCISSOR_TEST);
+    driver->glScissor(x, y, 1, 1);
+
+    driver->glClearStencil(0);
+
+    driver->glEnable(eGL_DEPTH_TEST);
+    driver->glDepthFunc(eGL_ALWAYS);
+    driver->glDepthMask(GL_TRUE);
+    driver->glDisable(eGL_BLEND);
+
+    driver->glStencilMask(0xff);
+    driver->glStencilOp(eGL_INCR, eGL_INCR, eGL_INCR);
+    driver->glEnable(eGL_STENCIL_TEST);
+
+    PixelModification referenceHistory;
+    referenceHistory.eventId = modEvents[i].eventId;
+
+    auto historyIndex =
+        std::lower_bound(history.begin(), history.end(), referenceHistory,
+                         [](const PixelModification &h1, const PixelModification &h2) -> bool {
+                           return h1.eventId < h2.eventId;
+                         });
+
+    RDCASSERT(historyIndex != history.end());
+
+    GLint savedReadFramebuffer, savedDrawFramebuffer;
+    driver->glGetIntegerv(eGL_DRAW_FRAMEBUFFER_BINDING, &savedDrawFramebuffer);
+    driver->glGetIntegerv(eGL_READ_FRAMEBUFFER_BINDING, &savedReadFramebuffer);
+
+    for(size_t j = 0; j < std::max(numFragments, 1u); ++j)
+    {
+      //  Set the stencil function so only jth fragment will pass.
+      driver->glStencilFunc(eGL_EQUAL, (int)j, 0xff);
+      driver->glClear(eGL_STENCIL_BUFFER_BIT);
+
+      driver->ReplayLog(modEvents[i].eventId, modEvents[i].eventId, eReplay_OnlyDraw);
```

Comment:
This has the same kind of problem as above - shader output values must be fetched direct to RGBA32F, otherwise if the original capture's color target is RGBA8 then we can't tell the difference between an output of 0.999 and 0.998 because they will be rounded to 1.0.

Note this is in contrast to premod/postmod which are _deliberately_ at the target output precision. That way you can see what value was output and then what value was actually stored to the target.